llama.cpp

```diff
@@ -13183,6 +13197,227 @@ int32_t llama_model_apply_lora_from_file(const struct llama_model * model, const
     }
 }
 
+struct llama_control_vector {
+    struct ggml_context * ctx;
+    std::vector<ggml_tensor*> tensors;
+
+    llama_control_vector() : ctx(nullptr) {}
+
+    ~llama_control_vector() {
+        if (this->ctx) {
+            ggml_free(this->ctx);
+        }
+    }
+};
+
+ggml_tensor * get_control_vector_layer_tensor(struct llama_control_vector * vector, int il) {
+    if (!vector->ctx || il > vector->tensors.size()) {
+        return nullptr;
+    }
+    return vector->tensors[il];
+}
+
+struct llama_control_vector * llama_control_vector_load(const char * path) {
+    struct llama_control_vector * vector = new llama_control_vector();
+
+    int n_tensors;
+    size_t n_bytes = 0;
+    uint32_t max_direction_layer = 0;
+
+    // calculate size of ctx needed for tensors, ensure tensors are f32, and find max layer
+    {
+        struct ggml_init_params meta_params = {
+            /* .mem_size   = */ ggml_tensor_overhead() * 128 + ggml_graph_overhead(),
+            /* .mem_buffer = */ nullptr,
+            /* .no_alloc   = */ true,
+        };
+        ggml_context * meta_ctx = ggml_init(meta_params);
+        struct gguf_init_params meta_gguf_params = {
+            /* .no_alloc = */ true,
+            /* .ctx      = */ &meta_ctx,
+        };
+        struct gguf_context * meta_ctx_gguf = gguf_init_from_file(path, meta_gguf_params);
+        if (!meta_ctx_gguf) {
+            LLAMA_LOG_ERROR("%s: failed to load control vector\n", __func__);
+            ggml_free(meta_ctx);
+            return nullptr;
+        }
+
+        n_tensors = gguf_get_n_tensors(meta_ctx_gguf);
+        for (int i = 0; i < n_tensors; i++) {
+            std::string name = gguf_get_tensor_name(meta_ctx_gguf, i);
+
+            // split on '.'
+            size_t dotpos = name.find('.');
```

Comment:
@ggerganov I notice that in llama.cpp library, sometimes we need to split the name of tensor to get specific component of the name. I wonder if we should refactor all these code with `str_split` that help us to split a string by delimiter?

https://github.com/ggerganov/llama.cpp/pull/5741/files#diff-e67669afc7d2ce9249080bc9118cdd58db64fd041f90cf98aa25aea7e82ac247R28